package com.amazon.problems;


class ListNode {
	 		int val;
	 		ListNode next;
	 		ListNode() {}
	 		ListNode(int val) { this.val = val; }
	 		ListNode(int val, ListNode next) { this.val = val; this.next = next; }
	  }



class LinkedList{
	ListNode head = null;
	
	public ListNode add(int data) {
		ListNode newNode = new ListNode(data);
		newNode.next = head;
		head = newNode;
		return head;
	}
	
	public void print(ListNode h1) {
		ListNode curr = h1;
		while(curr != null) {
			System.out.println(curr.val);
			curr = curr.next;
		}
	}
}

public class LinkedListProblems {
	
	// reverse list
	public static ListNode reverseList(ListNode head) {
		ListNode current = head;
		ListNode prev = null;
		ListNode next = null;
		while(current != null) {
			next = current.next;
			current.next = prev;
			prev = current;
			current = next;
		}
		head = prev;
		return head;
	}
	
	// odd-even node
	public static ListNode oddEvenList(ListNode head) {
		if(head == null || head.next == null) {return head;}
		ListNode odd = head;
		ListNode even = head.next;
		ListNode evenHead = even;
		while(even != null && even.next != null) {
			odd.next = even.next;
			odd = odd.next;
			even.next = even.next.next;
			even = even.next;
		}
		odd.next = evenHead;
		return head;
	}
	
	// detect cycle
	public static boolean hasCycle(ListNode head) {
		if(head == null || head.next == null) {
			return false;
		}
		ListNode slow = head;
		ListNode fast = head;
		while(fast != null && fast.next != null) {
			slow = slow.next;
			fast = fast.next.next;
			
			if(slow == fast) {
				return true;
			}
		}
		return false;
	}
	
	// detect cycle
	public static ListNode detectCycle(ListNode head) {
		if(head == null || head.next == null) {return null;}
		ListNode slow = head; 
		ListNode fast = head;
		while(fast != null && fast.next != null) {
			slow = slow.next;
			fast = fast.next.next;
			if(slow == fast) {
				ListNode ptr = head;
				while(ptr != slow) {
					ptr = ptr.next;
					slow = slow.next;
				}
				return ptr;
			}
		}
		return null;
	}
	
	// break cycle
	public static void breakCycle(ListNode head) {
		if(head==null || head.next == null) return;
		ListNode slow = head;
		ListNode fast = head;
		while(fast != null && fast.next != null) {
			slow = slow.next;
			fast = fast.next.next;
			if(slow == fast) {
				ListNode ptr = head;
				while(ptr != slow) {
					ptr = ptr.next;
					slow = slow.next;
				}
				ListNode temp = ptr;
				while(temp.next != ptr) {
					temp = temp.next;
				}
				temp.next = null;
				return;
			}
		}
	}
	
	public static void main(String[] args) {
		ListNode n1 = new ListNode(1);
		ListNode n2 = new ListNode(2);
		ListNode n3 = new ListNode(3);
		ListNode n4 = new ListNode(4);
		ListNode n5 = new ListNode(5);
		
		ListNode head = n1;
		n1.next = n2;
		n2.next = n3;
		n3.next = n4;
		n4.next = n5;
		
		ListNode current = head;
		while(current != null) {
			current = current.next;
		}
		ListNode newHead = oddEvenList(head);
		current = newHead;
		while(current != null && current.next != null) {
			current = current.next;
		}
	}

}
